express-notes-api/index.js

```diff
@@ -0,0 +1,116 @@
+const express = require('express');
+const app = express();
+const fs = require('fs');
+app.use(express.json());
+
+const data = require('./data.json');
+const notes = data.notes;
+
+app.get('/api/notes', (req, res) => {
+  const notesArr = [];
+  for (const item in notes) {
+    notesArr.push(notes[item]);
+  }
+  res.send(notesArr);
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const parsedId = parseInt(req.params.id, 10);
+  if (notes[parsedId] !== undefined) {
+    res.send(notes[parsedId]);
+  } else if (isNaN(parsedId)) {
+    res.status(400).send({
+      error: 'id must be a positive integer'
+    });
+  } else {
+    res.status(404).send({
+      error: `cannot find note with id ${parsedId}`
+    });
+  }
+});
+
+app.post('/api/notes', (req, res) => {
+  if (req.body.content !== undefined) {
+    const newNote = {
+      id: data.nextId,
+      content: req.body.content
+    };
+    notes[data.nextId] = newNote;
+    data.nextId++;
+    const dataStr = JSON.stringify(data, null, 2);
+    fs.writeFile('data.json', dataStr, 'utf8', err => {
+      if (err) {
+        res.status(505).send({
+          error: 'An unexpected error occurred.'
+        });
+      }
+      res.status(201).json(newNote);
```

Comment:
Be sure to include either a `return` statement or an `else` block to prevent the server from attempting to respond twice. 1 Request -> 1 Response. `res.send()` does not end the function, only `return` and `throw` do that in JavaScript.

Reply:
This should also be addressed elsewhere in your code.